---
title: "R-assignment"
author: "Tianqi(Tim)"
date: "10/12/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown

This is an R Markdown document. Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents. For more details on using R Markdown see <http://rmarkdown.rstudio.com>.

When you click the **Knit** button a document will be generated that includes both content as well as the output of any embedded R code chunks within the document. You can embed an R code chunk like this:

```{r}
#part1
if (!require("tidyverse")) install.packages("tidyverse")
library(tidyverse)
fang_et_al_genotypes <- read_tsv ("https://raw.githubusercontent.com/Kakashi-sensei/Tianqili-R-Assignment/master/fang_et_al_genotypes.txt")
snp_position <- read_tsv ("https://raw.githubusercontent.com/Kakashi-sensei/Tianqili-R-Assignment/master/snp_position.txt")
maize_fang_et_al_genotypes <- filter(fang_et_al_genotypes, Group == "ZMMLR" | Group == "ZMMIL" | Group == "ZMMMR")
teosinte_fang_et_al_genotypes <- filter(fang_et_al_genotypes, Group == "ZMPBA" | Group == "ZMPIL" | Group == "ZMPJA")
maize_fang_et_al_genotypes <- maize_fang_et_al_genotypes[,-c(1,2,3)]
maize_fang_et_al_genotypes <- t (maize_fang_et_al_genotypes)
teosinte_fang_et_al_genotypes <- teosinte_fang_et_al_genotypes[,-c(1,2,3)]
teosinte_fang_et_al_genotypes <- t (teosinte_fang_et_al_genotypes)
snp_position <- snp_position[,-c(2,5:15)]
arrange(snp_position, SNP_ID)
maize_fang_et_al_genotypes <- cbind(snp_position, maize_fang_et_al_genotypes)
teosinte_fang_et_al_genotypes <- cbind(snp_position, teosinte_fang_et_al_genotypes)
#replace "?" to "-"
maize_ded <- data.frame(lapply(maize_fang_et_al_genotypes, as.character), stringsAsFactors=FALSE)
maize_ded <- data.frame(sapply(
  maize_ded,function(x) {
    x <- gsub("?","-",x,fixed=TRUE)
                        }
                             ))
teosinte_ded <- data.frame(lapply(teosinte_fang_et_al_genotypes, as.character), stringsAsFactors=FALSE)
teosinte_ded <- data.frame(sapply
                            (teosinte_ded,function(x){
                              x <- gsub("?","-",x,fixed=TRUE)
                                                     }
                              ))
rm(fang_et_al_genotypes)#clear up
rm(snp_position)
#game on
for (i in 1:10) {
  maize <- filter(maize_fang_et_al_genotypes, Chromosome == i)
  maize <- arrange(maize, Position)
  outpath <- "c:/Users/46554/github/R-Assignment/Tianqili-R-Assignment/submit/"
  nam <- sapply(
    names(maize),function(x){
                    paste("maize_in", i, ".csv", sep='')
                            }
               )
  out_filePath <- sapply(
    nam, function(x){
             paste(outpath, x, sep='/')
                    }
                         )
  write.csv(maize, file=out_filePath[i])
                } #10 of maize increase data
for (i in 1:10) {
  teosinte <- filter(teosinte_fang_et_al_genotypes, Chromosome == i)
  teosinte <- arrange(teosinte, Position)
  outpath <- "c:/Users/46554/github/R-Assignment/Tianqili-R-Assignment/submit/"
  nam <- sapply(
    names(teosinte),function(x){
                    paste("teosinte_in", i, ".csv", sep='')
                               }
               )
  out_filePath <- sapply(
                    nam, function(x){
                      paste(outpath, x, sep='/')
                                    }
                         )
   write.csv(teosinte, file=out_filePath[i])
                } #10 of teosinte increase data
for (i in 1:10) {
  maize <- filter(maize_ded, Chromosome == i)
  maize <- arrange(maize, desc(Position))
  outpath <- "c:/Users/46554/github/R-Assignment/Tianqili-R-Assignment/submit/"
  nam <- sapply(
    names(maize),function(x){
                    paste("maize_de", i, ".csv", sep='')
                            }
               )
  out_filePath <- sapply(nam, function(x){
                     paste(outpath, x, sep='/')
                                         }
                        )
  write.csv(maize, file=out_filePath[i])
                 }
for (i in 1:10) {
  teosinte <- filter(teosinte_ded, Chromosome == i)
  teosinte <- arrange(teosinte, desc(Position))
  outpath <- "c:/Users/46554/github/R-Assignment/Tianqili-R-Assignment/submit/"
  nam <- sapply(
    names(teosinte),function(x){
                    paste("teosinte_de", i, ".csv", sep='')
                               }
                )
  out_filePath <- sapply(
    nam, function(x){
                     paste(outpath, x, sep='/')
                    }
                        )
  write.csv(teosinte, file=out_filePath[i])
                 }
rm(list = ls())
```


```{r}
#Part2
##SNPs per chromosome
if (!require("tidyverse")) install.packages("tidyverse")
library(tidyverse)
if (!require("reshape2")) install.packages("reshape2")
library(reshape2)
fang_et_al_genotypes <- read_tsv ("https://raw.githubusercontent.com/Kakashi-sensei/Tianqili-R-Assignment/master/fang_et_al_genotypes.txt")
snp_position <- read_tsv ("https://raw.githubusercontent.com/Kakashi-sensei/Tianqili-R-Assignment/master/snp_position.txt")
maize_fang_et_al_genotypes <- filter(fang_et_al_genotypes, Group == "ZMMLR" | Group == "ZMMIL" | Group == "ZMMMR")
teosinte_fang_et_al_genotypes <- filter(fang_et_al_genotypes, Group == "ZMPBA" | Group == "ZMPIL" | Group == "ZMPJA")
maize_fang_et_al_genotypes <- maize_fang_et_al_genotypes[,-c(1,2,3)]#cut head
maize_fang_et_al_genotypes <- t (maize_fang_et_al_genotypes)
teosinte_fang_et_al_genotypes <- teosinte_fang_et_al_genotypes[,-c(1,2,3)]#Guillotine
teosinte_fang_et_al_genotypes <- t (teosinte_fang_et_al_genotypes)
snp_position <- snp_position[,-c(2,5:15)]#Guillotine
arrange(snp_position, SNP_ID)
maize_fang_et_al_genotypes <- cbind(snp_position, maize_fang_et_al_genotypes)
teosinte_fang_et_al_genotypes <- cbind(snp_position, teosinte_fang_et_al_genotypes)
rm(fang_et_al_genotypes)#clear up
rm(snp_position)#clear up
setwd("c:/Users/46554/github/R-Assignment/Tianqili-R-Assignment/submit/")#set saving address
i <- c(1, 2)#Plot the total number of SNPs in our dataset on each chromosome
  if(i==1){
    pp=ggplot(data = maize_fang_et_al_genotypes) + 
      geom_bar(mapping = aes(x = Chromosome))
    print(pp)
    ggsave(pp,filename = "maize_p.pdf",width = 3,height = 3)
    dev.off()
    getwd()#save files
          }else{
            pp=ggplot(data = teosinte_fang_et_al_genotypes) + 
             geom_bar(mapping = aes(x = Chromosome))
            print(pp)
            ggsave(pp,filename = "teosinte_p.pdf",width = 3,height = 3)
            dev.off()#save files
            getwd()
            }
    rm(i)
    rm(pp)
i <- c(1, 2)#plot the distribution(position) of SNPs on chromosomes
  maize_fang_et_al_genotypes$Position <- as.numeric(maize_fang_et_al_genotypes$Position)
  teosinte_fang_et_al_genotypes$Position <- as.numeric(teosinte_fang_et_al_genotypes$Position)
  if(i==1){
  pp=ggplot(data = maize_fang_et_al_genotypes, mapping=aes(x=Chromosome, y=Position))+
      geom_point(alpha=0.1)
  print(pp)
  ggsave(pp,filename = "maize_distribution.pdf",width = 3,height = 3)
  dev.off()
  getwd()
          }else{
  pp=ggplot(data = teosinte_fang_et_al_genotypes, mapping=aes(x=Chromosome, y=Position))+
     geom_point(alpha=0.1)
  print(pp)
  ggsave(pp,filename = "teosinte_distribution.pdf",width = 3,height = 3)
  dev.off()
  getwd()
               }
rm(pp)
rm(i)
##Missing data and amount of heterozygosity
i <- c(1, 2)
  if(i==1){
    maize_melt <-melt(maize_fang_et_al_genotypes,id.vars=c("SNP_ID","Chromosome", "Position"),variable.name = "Sample", value.name="Nucleotide")
    rm(maize_fang_et_al_genotypes)
nu <- data.frame(lapply(maize_melt, as.character), stringsAsFactors=FALSE)
nu <- data.frame(lapply(nu,function(x) {x <- gsub("A","1",x,fixed=TRUE)}))
nu <- data.frame(lapply(nu,function(x) {x <- gsub("C","2",x,fixed=TRUE)}))
nu <- data.frame(lapply(nu,function(x) {x <- gsub("G","3",x,fixed=TRUE)}))
nu <- data.frame(lapply(nu,function(x) {x <- gsub("T","4",x,fixed=TRUE)}))
nu <- data.frame(lapply(nu,function(x) {x <- gsub("?","0",x,fixed=TRUE)}))
nu$Nucleotide <- c(nu$Nucleotide)
maize_melt <- mutate(maize_melt, allele = case_when(nu$Nucleotide==2~ "homozygous",
                                                    nu$Nucleotide==11~ "homozygous",
                                                    nu$Nucleotide==9~ "homozygous",
                                                    nu$Nucleotide==6~ "homozygous",
                                                    nu$Nucleotide==1~ "miss",
                                                    nu$Nucleotide!=(1|2|11|9|6)~ "heterozygous")
                    )
           }else{ 
teosinte_melt<-melt(teosinte_fang_et_al_genotypes,id.vars=c("SNP_ID","Chromosome", "Position"),variable.name = "Sample", value.name="Nucleotide")
rm(teosinte_fang_et_al_genotypes)
nu <- data.frame(lapply(teosinte_melt, as.character), stringsAsFactors=FALSE)
nu <- data.frame(lapply(nu,function(x) {x <- gsub("A","1",x,fixed=TRUE)}))
nu <- data.frame(lapply(nu,function(x) {x <- gsub("C","2",x,fixed=TRUE)}))
nu <- data.frame(lapply(nu,function(x) {x <- gsub("G","3",x,fixed=TRUE)}))
nu <- data.frame(lapply(nu,function(x) {x <- gsub("T","4",x,fixed=TRUE)}))
nu <- data.frame(lapply(nu,function(x) {x <- gsub("?","0",x,fixed=TRUE)}))
nu$Nucleotide <- c(nu$Nucleotide)
teosinte_melt <- mutate(teosinte_melt, allele = case_when(nu$Nucleotide==2~ "homozygous",
                                                          nu$Nucleotide==11~ "homozygous",
                                                          nu$Nucleotide==9~ "homozygous",
                                                          nu$Nucleotide==6~ "homozygous",
                                                          nu$Nucleotide==1~ "miss",
                                                          nu$Nucleotide!=(1|2|11|9|6)~ "heterozygous")
                        )
                 }
rm(nu)
rm(i)
#proportion of homozygous and heterozygous sites as well as missing data in each sample

i<- c(1,2)
  if(i==1){
pp=ggplot(data = maize_melt) + 
  geom_bar(mapping = aes(x =Sample, fill=allele))
print(pp)
ggsave(pp,filename = "maize_proportion.pdf",width = 3,height = 3)
dev.off()
getwd()
          }else{
            pp=ggplot(data = teosinte_melt) + 
            geom_bar(mapping = aes(x =Sample, fill=allele))
            print(pp)
            ggsave(pp,filename = "teosinte_proportion.pdf",width = 3,height = 3)
            dev.off()
            getwd()
                }
  rm(pp)
  rm(i)
#proportion of homozygous and heterozygous sites as well as missing data in each Chromosome group
setwd("c:/Users/46554/github/R-Assignment/Tianqili-R-Assignment/submit/")
 i<- c(1, 2)
  if(i==1){
pp=ggplot(data = maize_melt) + 
  geom_bar(mapping = aes(x =Chromosome, fill=allele), position = 'fill')#Normalize the height of individual bars
print(pp)
ggsave(pp,filename = "maize_gro_proportion.pdf",width = 3,height = 3)
dev.off()
getwd()
          }else{
            pp=ggplot(data = teosinte_melt) + 
            geom_bar(mapping = aes(x =Chromosome, fill=allele), position = 'fill')#Normalize the height of individual bars
            print(pp)
            ggsave(pp,filename = "teosinte_gro_proportion.pdf",width = 3,height = 3)
            dev.off()
            getwd()
                }
  rm(pp)
  rm(i)
#My own visualization-proportion of homozygous and heterozygous sites as well as missing data of all samples
i <- c(1, 2)
  if(i==1){
pp=ggplot(data = maize_melt) + 
  geom_bar(mapping = aes(x =allele))
print(pp)
ggsave(pp,filename = "maize_allele.pdf",width = 3,height = 3)
dev.off()
getwd()
          }else{
            pp=ggplot(data = teosinte_melt) + 
            geom_bar(mapping = aes(x =allele))
            print(pp)
            ggsave(pp,filename = "teosinte_allele.pdf",width = 3,height = 3)
            dev.off()
            getwd()
                }
rm(list = ls())
```

